[refactor] Config handling refactoring in preparation for network-specific sections

ajtowns opened the pull request:
This does a bit of refactoring of the configuration handling code in order to add additional tests to make adding support for [test]/[regtest] sections in the config file in #11862 easier. Should not cause any behaviour changes.

ajtowns committed:
Move ChainNameFromCommandLine into ArgsManager and rename to GetChainName

ajtowns committed:
[tests] Add unit tests for GetChainName

ajtowns committed:
Separate out ReadConfigStream from ReadConfigFile

ajtowns committed:
ReadConfigStream: assume the stream is good

ajtowns committed:
[tests] Add unit tests for ReadConfigStream

ajtowns committed:
[tests] Check GetChainName works with config entries

ajtowns committed:
[tests] Add additional unit tests for -nofoo edge cases

jnewbery commented on src/test/util_tests.cpp:
```diff
@@ -253,16 +258,150 @@ BOOST_AUTO_TEST_CASE(util_GetBoolArgEdgeCases)
+    BOOST_CHECK(test_args.GetArgs("-fff").size() == 1
+                && test_args.GetArgs("-fff").front() == "0");
+    BOOST_CHECK(test_args.GetArgs("-nofff").size() == 0);
+    BOOST_CHECK(test_args.GetArgs("-h").size() == 2
```
Why no: `    BOOST_CHECK(test_args.GetArgs("-ggg").size() == 1);` ?

jnewbery commented on src/test/util_tests.cpp:
```diff
@@ -253,16 +258,150 @@ BOOST_AUTO_TEST_CASE(util_GetBoolArgEdgeCases)
+    BOOST_CHECK(!test_args.IsArgNegated("-ccc"));
+    BOOST_CHECK(!test_args.IsArgNegated("-d"));
+    BOOST_CHECK(test_args.IsArgNegated("-fff"));
+    BOOST_CHECK(test_args.IsArgNegated("-ggg")); // IsArgNegated==true when noggg=0
```
> IsArgNegated==true when noggg=0

yuck. Can we change that?

ajtowns replied:
It's changed in #11862 (and easy to change on its own as well). I think it makes sense for this PR to just refactor/add tests rather than change functionality though.

jnewbery replied:
sounds good. Keep this PR as refactor only and do behaviour changes in #11862.

jnewbery reviewed (commented):
Looks really good. Thanks for all the additional test coverage!

ajtowns commented:
Bumped to address @jnewbery's suggestions. Unsquashed changes are at https://github.com/ajtowns/bitcoin/commits/netconf-refactor-track with commit 26d3faaa28c794b48df2a12cdb4a9a71de3d1a96 here having the same tree  as commit 3467454beff8e7b2df87f9a0e29391eb14a8e6c3 on the -track branch:

```
$ for a in netconf-refactor{,-track}; do git log -1 --pretty="%h -> %T" $a; done
26d3faaa28 -> ef5b209a05c28afb98e0ce8e05e9cbc54dfebeaf
3467454bef -> ef5b209a05c28afb98e0ce8e05e9cbc54dfebeaf
```

Empact commented on src/util.cpp:
```diff
@@ -735,35 +735,55 @@ fs::path GetConfigFile(const std::string& confPath)
     }
 }
 
+std::string ArgsManager::ChainNameFromCommandLine() const
```
Could just call this `ChainName` given `gArgs` is all about the command line / config files already.

ajtowns replied:
:+1:  Probably GetChainName() would be slightly more consistent...

jnewbery replied:
Agree that `GetChainName()` is a better name. Can be done in this PR or a future PR.

jnewbery commented:
utACK 26d3faaa28c794b48df2a12cdb4a9a71de3d1a96

ajtowns commented:
Rebased due to conflict with #10244, and renamed ChainNameFromCommandLine to GetChainName. Not sure how useful the -track branch is since it's now got a merge with master, but it still works: https://github.com/ajtowns/bitcoin/commit/492bbcdc558ae0edae0e2beb49a50932c8d576c6 follows on from where the tracking branch left off, and matches this branches current head, 77a733a99a75bdd04ad94df830e5bdc8a6040959:

```
$ for a in netconf-refactor{,-track}; do git log -1 --pretty="%h -> %T" $a; done
77a733a99a -> 4b6b60ed983260f4f3219ad28825d209ff5a7957
492bbcdc55 -> 4b6b60ed983260f4f3219ad28825d209ff5a7957
```

jtimon commented:
utACK 77a733a99a75bdd04ad94df830e5bdc8a6040959

theuni commented on src/util.cpp:
```diff
@@ -737,31 +737,31 @@ fs::path GetConfigFile(const std::string& confPath)
 
 void ArgsManager::ReadConfigStream(std::istream& stream)
 {
-    if (!stream.good())
```
Replace this with an assertion instead, even if just to self-document assumptions?

theuni commented:
utACK 77a733a99a75bdd04ad94df830e5bdc8a6040959, not too concerned about the nit.

jnewbery commented:
ACK 77a733a99a75bdd04ad94df830e5bdc8a6040959

MeshCollider commented:
Concept ACK, this should also help my arg refactor a bit

jimpo commented:
utACK 77a733a

MarcoFalke commented on src/test/util_tests.cpp:
```diff
@@ -258,16 +258,52 @@ BOOST_AUTO_TEST_CASE(util_GetBoolArgEdgeCases)
 
     // A double negative is a positive.
     BOOST_CHECK(testArgs.IsArgNegated("-bar"));
-    BOOST_CHECK(testArgs.GetBoolArg("-bar", false) == true);
```
In the test case `util_GetBoolArgEdgeCases` you seem to be removing all calls to `GetBoolArg`. Is this intentional?

ajtowns replied:
Yeah; I think the edge cases are more clearly tested with GetArg and a clearly distinct default value. The IsArgNegated/GetBoolArg cases are still tested in util_GetBoolArg (via -a and -b).

MarcoFalke commented:
post merge utACK 77a733a beside question
